### ISSUE TITLE
topology: merge eq_fir and eq_iir IPC constructors into generic "effect" constructor. 

### ISSUE THREAD

**Issue by lgirdwood** [thesofproject/linux#618]
These can be merged and reused for voice detection. All the differentiation is done in the private data.
@plbossart fyi, this is what I'm thinking about changing this weekend. Seems we have unnecessary duplication for no benefit (which will lead to more duplication for other new "effects").
```
/* FIR equalizer component */
struct sof_ipc_comp_eq_fir {
	struct sof_ipc_comp comp;
	struct sof_ipc_comp_config config;
	uint32_t size;

	/* reserved for future use */
	uint32_t reserved[8];

	unsigned char data[0];
} __packed;

/* IIR equalizer component */
struct sof_ipc_comp_eq_iir {
	struct sof_ipc_comp comp;
	struct sof_ipc_comp_config config;
	uint32_t size;

	/* reserved for future use */
	uint32_t reserved[8];

	unsigned char data[0];
} __packed;
```

**Comment by plbossart** [comment]
On 2/1/19 4:02 PM, Liam Girdwood wrote:
>
> These can be merged and reused for voice detection. All the 
> differentiation is done in the private data.
> @plbossart <https://github.com/plbossart> fyi, this is what I'm 
> thinking about changing this weekend. Seems we have unnecessary 
> duplication for no benefit (which will lead to more duplication for 
> other new "effects").
>
> |/* FIR equalizer component */ struct sof_ipc_comp_eq_fir { struct 
> sof_ipc_comp comp; struct sof_ipc_comp_config config; uint32_t size; 
> /* reserved for future use */ uint32_t reserved[8]; unsigned char 
> data[0]; } __packed; /* IIR equalizer component */ struct 
> sof_ipc_comp_eq_iir { struct sof_ipc_comp comp; struct 
> sof_ipc_comp_config config; uint32_t size; /* reserved for future use 
> */ uint32_t reserved[8]; unsigned char data[0]; } __packed; |
"effect" usually means PCM in and out. If you deal with 
non-audio/compressed data you are really talking about a processing 
element using non-descript binary data - or a payload. I would remove 
any type here, just pass the data and let the component parse it.

**Comment by singalsu** [comment]
The sof_ipc_comp_process has replaced in firmware the processing algorithm specific IPC structs.

```
/* generic "effect", "codec" or proprietary processing component */
struct sof_ipc_comp_process {
	struct sof_ipc_comp comp;
	struct sof_ipc_comp_config config;
	uint32_t size;	/**< size of bespoke data section in bytes */
	uint32_t type;	/**< sof_ipc_effect_type */

	/* reserved for future use */
	uint32_t reserved[7];

	unsigned char data[0];
} __attribute__((packed));

```

The same change still need to be done on Linux driver side as well. It is in Linux PR #771 by @keyonjie .

**Comment by singalsu** [comment]
This can be closed since #771 is merged now.
